src/main/typescript/rug/operations/Decorators.ts

```diff
@@ -27,137 +30,190 @@ function get_metadata(obj: any, key: string) {
-    }
+  expression?: PathExpression<any, any> | string) {
+  return (obj: object) => {
     if (expression !== undefined) {
-      set_metadata(ctr.prototype, "__name", nameOrDescription);
-      set_metadata(ctr.prototype, "__description", descriptionOrExpression);
-      set_metadata(ctr.prototype, "__kind", "event-handler");
-      if (typeof expression === "string") {
-        set_metadata(ctr.prototype, "__expression", expression);
-      } else {
-        set_metadata(ctr.prototype, "__expression", expression.expression);
-      }
-    } else {
-      set_metadata(ctr.prototype, "__description", nameOrDescription);
-      set_metadata(ctr.prototype, "__kind", "event-handler");
       if (typeof descriptionOrExpression === "string") {
-        set_metadata(ctr.prototype, "__expression", descriptionOrExpression);
+        declareEventHandler(obj, descriptionOrExpression, expression, nameOrDescription);
       } else {
-        set_metadata(ctr.prototype, "__expression", descriptionOrExpression.expression);
+        throw Error("When three parameters are passed, the second must be a description");
       }
+    } else {
+      declareEventHandler(obj, nameOrDescription, descriptionOrExpression);
     }
   };
-};
+}
+
+export function declareEventHandler(
+  obj: any, description: string, expression: string | PathExpression<any, any>, name?: string) {
+  declareRug(obj, "event-handler", description, name);
+  if (typeof expression === "string") {
+    set_metadata(obj, "__expression", expression);
+  } else {
+    set_metadata(obj, "__expression", expression.expression);
+  }
+  return obj;
+}
 
 /**
  * Decorator for tags. Sets tags on the class
  */
 
-function Tags(...tags: string[]) {
+export function Tags(...tags: string[]) {
   return (target: any) => {
-    set_metadata(target.prototype, "__tags", tags);
+    declareTags(target, tags);
   };
 }
 
-function Intent(...intent: string[]) {
+export function declareTags(target: any, tags: string[]) {
+  set_metadata(target, "__tags", tags);
+  return target;
+}
+
+export function Intent(...intent: string[]) {
   return (target: any) => {
-    set_metadata(target.prototype, "__intent", intent);
+    declareIntent(target, intent);
   };
 }
 
-function Secrets(...secrets: string[]) {
+export function declareIntent(target: any, intent: string[]) {
+  set_metadata(target, "__intent", intent);
+  return target;
+}
+
+export function Secrets(...secrets: string[]) {
   return (target: any) => {
-    let current = get_metadata(target.prototype, "__secrets");
-    if (current == null) {
-      current = [];
-    }
-    set_metadata(target.prototype, "__secrets", current.concat(secrets));
+    declareSecrets(target, secrets);
   };
 }
 
+export function declareSecrets(target: any, secrets: string[]) {
+  let current = get_metadata(target.prototype, "__secrets");
+  if (current == null) {
+    current = [];
+  }
+  set_metadata(target, "__secrets", current.concat(secrets));
```

Comment:
I think not. The reason we use concat is to create a copy, not so that it can be called multiple times. As above, properties defined on classes via decorators are immutable.